design/baremetal-operator/raid-api.md

```diff
@@ -0,0 +1,198 @@
+# raid-api
+
+## Status
+
+implementable
+
+## Summary
+
+This design describes changes needed to allow RAID storage
+configuration as part of provisioning hosts.
+
+## Motivation
+
+While it is reasonable to assume that cloud instances and VMs only
+need one drive with minimal configuration, many classes of server
+hardware need to have RAID storage configured to present fast and
+reliable storage to the OS. While it is possible today to
+pre-configure the hosts, it would be more convenient for users is
+metal3 could do that work for them.
+
+### Goals
+
+1. Provide an API to specify the RAID storage configuration to use
+   when provisioning a host.
+2. Outline the controller/provisioner logic changes needed to pass the
+   instructions to Ironic to have the storage configured.
+3. Support basic RAID levels `0`, `1`, and `1+0`.
+
+### Non-Goals
+
+1. Provide default RAID configuration.
+2. Retain existing RAID configuration on a host if it matches the
+   desired configuration at the time of provisioning (we make no
+   assumptions about preserving the state of the storage system or its
+   contents).
+3. Re-provision a host if the RAID settings change. (Users are
+   expected to explicitly deprovision the host and then reprovision
+   it.)
+4. Support every possible RAID configuration.
+
+## Proposal
+
+### User Stories
+
+1. As a user, I can specify the RAID configuration for a host before
+   provisioning it.
+2. As a user, I can choose between hardware and software RAID for each
+   host.
+
+## Design Details
+
+### API Changes
+
+The `BareMetalHost` API will change to accept RAID configuration
+inputs and to reflect the settings that were used when a host is
+provisioned (in case they are changed after provisioning).
+
+New `spec` fields will be added in a section called `raid`, with
+separate fields describing the hardware and software RAID settings. If
+the hardware settings are provided, the software settings are ignored.
+
+For hardware RAID, the inputs describe how to combine physical disks
+into a RAID volume via the RAID controller hardware in the host.
+
+```yaml
+spec:
+  raid:
+    hardwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        name: volume-name
+        rotational: true
+        numberOfPhysicalDisks: 2
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The number of disks to use for the volume defaults to a minimum
+  value based on the RAID level.
```

Comment:
Hardware RAID offered by ironic hardware types/drivers may have additional requirements. For example, [here](https://github.com/openstack/ironic/blob/72baa1da581e42042e488abbdb293e5f9f26b9a2/ironic/drivers/modules/drac/raid.py#L1233-L1241) the `idrac` hardware type requires that the physical disks be specified when size is set to `MAX`.

Reply:
Yes, but not all BMCs require this, and it is difficult to make BMO consistent with "ironic" in a timely manner. So can we let `ironic` check this?

Reply:
Yes, but not all BMCs require this, and it is difficult to make BMO consistent with "ironic" in a timely manner. So can we let ironic check this?

Reply:
We could add this to our ``idrac`` implementation as a small check in the code. Yes?

Reply:
If we can anticipate these sorts of issues in metal3, that would be good. Otherwise it may take quite a while between the user providing the RAID settings and Ironic reporting an error.

It's also OK, IMHO, for us to limit the configuration options metal3 supports to be a subset of what Ironic supports, if we're focusing on making the hosts suitable to be nodes in a cluster.

So, we could do something like default to listing all of the physical disks when the size is MAX, even if that is not the default behavior of Ironic.

Reply:
That's actually a very reasonable and good approach, since after all we're breaking fresh ground here, implementing RAID support. We can always extend this to encase all of what Ironic can offer us.